docs/proposals/007-variadic-generics.md

```diff
@@ -0,0 +1,662 @@
+
+```
+// The callback function `p` is represented as a functor conforming to the `IProcessor` interface.
+void process<TProcessor:IFunc<void, int>>(TProcessor p, int data[N])
+{
+    for (int i = 0; i < N; i++)
+        p.process(data[i]);
+}
+
+void user(int myData[N])
+{
+    process((int x)=>{...}, myData);
+}
+```
+
+Related Work
+------------
+
+Variadic generics is an advance type system feature that is missinng in many modern languages including C# and Rust.
+Swift adds support for variadic generics recently in late 2022/2023, and this proposal largely follows Swift's design.
+C++ has variadic templates that achieves similar results within the template system.
+Rust supports variadics in a macro system layered above its core type system. While this can solve many user issues,
+we decided to not go through this path because macros and templates must be expanded before core type checking, which means that
+they can't integrate nicely in modules and compiled into IR independently of their use sites.
+
+
+Proposed Approach
+-----------------
+
+Slang can follow Swift's solution for variadic generics. A user can define a variadic generic with the syntax:
+
+```
+void myFunc<each T>(expand each T v) {...}
+```
+
+The code above defines a generic function that has a __generic type pack parameter__ `T` with the `each` keyword before `T`.
+The function's parameter list is defined as `expand each T v`, which should be interpreted as a parameter `v` whose type is
+`expand each T`. `expand each T` is a type that represents a pack of types. A parameter whose type is a pack of types can
+accept zero or more arguments during function call resolution.
+
+`myFunc` can be called with arbitrary number of arguments:
+```
+myFunc(); // OK, zero arguments
+myFunc(1, 2.0f, 3.0h); // OK, three arguments with different types.
+```
+
+A function can forward its variadic parameter to another function that accepts variadic parameter with the `expand` expression:
+```
+void caller<each T>(expand each T v)
+{
+    myFunc(expand each v);
+}
+```
+
+Generic type pack parameters can be nested, and there can be more than one variadic generic parameters in a single generic decl:
+```
+struct Parent<each T>
+{
+    void f<each U>(...) {...} // OK, nested generics with type pack parameters
+}
+void g<each T, each U>(...) { ... } // OK, more than one type pack parameter in a single generic.
+```
+
+However, when more than one generic type pack parameters is referenced in a single `expand` expression, there is an implicit
+requirement that these type packs will have the same number of elements. For example:
+```
+ // implicitly requiring T and U to have same number of elements.
+void g<each T, each U>(expand Pair<each T, each U> pairs) {...}
+
+void user()
+{
+    // We will match (int, float) to `T`, and (uint16_t, half) to `U`:
+    g<int, float, uint16_t, half>(
+        Pair<int, uint16_t>(1, 2),
+        Pair<float, half>(1.0f, 2.0h) );
+}
+```
```

Comment:
This example is really hard to understand to me.

I can get the part that `T = (int, float)` and `U = (uint16_t, half)`

But I really don't understand why 
`expand Pair<each T, each U> pairs` can be subtituted
with `(Pair<int, uint16_t>, Pair<float, half>)`.

Because it seems like now `T=int` and `U=uint16_t` for the first pair.
and `T=float` and `U=half` for the second pair.

Reply:
That's correct, `expand X` where `X` contains `each T` / `each U` means that if T and U are equally sized packs of `n` elements, we are going to generate `(X1, X2, ... Xn)`, where `Xi` is `X` with `each T` replaced with `T[i]` and `each U` replaced with `U[i]`.

For example, given
```
typealias M<each T> = expand vector<each T, 3>;
```
Then `M<TypePack(int ,float)>` is:
`(vector<T[0], 3>, vector<T[1], 3>)`.
Substitute T[0] = int and T[1] = float, we will get:
```
(vector<int, 3>, vector<float, 3>)
```

Reply:
I'm not sure if I'm really understand here

For your example here, can we use it as this?
```
    g<int, float, uint16_t, half>(
        Pair<int>(1),
        Pair<uint16_t>(2),
        Pair<float>(1.0f),
        Pair<half>(2.0h));
```

And how about this?
```
    g<int, float, uint16_t, half>(
        Pair<int, uint16_t>(1, 2),
        Pair<int, uint16_t>(3, 4),
        Pair<float, half>(1.0f, 2.0h),
        Pair<float, half>(3.0f, 4.0h) );
```